src/main/java/it/aboutbits/springboot/emailservice/lib/application/ManageEmail.java

```diff
@@ -0,0 +1,150 @@
+
+
+import it.aboutbits.springboot.emailservice.lib.AttachmentDataSource;
+import it.aboutbits.springboot.emailservice.lib.AttachmentReference;
+import it.aboutbits.springboot.emailservice.lib.EmailDto;
+import it.aboutbits.springboot.emailservice.lib.EmailState;
+import it.aboutbits.springboot.emailservice.lib.exception.AttachmentException;
+import it.aboutbits.springboot.emailservice.lib.jpa.EmailRepository;
+import it.aboutbits.springboot.emailservice.lib.model.Email;
+import it.aboutbits.springboot.emailservice.lib.model.EmailAttachment;
+import jakarta.mail.MessagingException;
+import jakarta.validation.Valid;
+import lombok.NonNull;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.core.io.ByteArrayResource;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.mail.javamail.MimeMessageHelper;
+import org.springframework.validation.annotation.Validated;
+
+import java.io.IOException;
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+
+@Validated
+@Log4j2
+public class ManageEmail {
+    private final EmailRepository emailRepository;
+    private final JavaMailSender mailSender;
+    private final AttachmentDataSource attachmentDataSource;
+    private final EmailMapper emailMapper;
+
+    public ManageEmail(
+            EmailRepository emailRepository,
+            JavaMailSender mailSender,
+            AttachmentDataSource attachmentDataSource,
+            final EmailMapper emailMapper) {
+
+        this.emailRepository = emailRepository;
+        this.mailSender = mailSender;
+        this.attachmentDataSource = attachmentDataSource;
+        this.emailMapper = emailMapper;
+    }
+
+    public EmailDto schedule(@NonNull @Valid EmailParameter parameter) {
+        var emailData = parameter.email();
+
+        final var email = new Email();
+        email.setState(EmailState.PENDING);
+        email.setSendingScheduledAt(parameter.scheduleAt());
+        email.setReference(parameter.reference());
+        email.setSubject(emailData.subject());
+        email.setTextBody(emailData.textBody());
+        email.setHtmlBody(emailData.htmlBody());
+        email.setRecipients(emailData.recipients());
+        email.setFromAddress(emailData.fromAddress());
+        email.setFromName(emailData.fromName());
+        email.setAttachments(emailData.attachments().stream()
+                .map(a -> {
+                    var attachment = new EmailAttachment();
+                    attachment.setEmail(email);
+                    attachment.setReference(a.reference().value());
+                    attachment.setContentType(a.contentType());
+                    attachment.setFileName(a.fileName());
+                    return attachment;
+                })
+                .collect(Collectors.toSet())
+        );
+
+        var savedEmail = emailRepository.save(email);
+
+        return emailMapper.toDto(savedEmail);
+    }
+
```

Comment:
I would propose to add a method to send an email synchronously. That is sometimes very convenient, when you want the email to send immediately.

Reply:
in that case I would propose to add a `sendNowOrFail()` action. And I would have it throw a checked exception to disencourage the use of it unless there is some special reason you need everything to fail just because the email had some issue.

Reply:
Sounds good. I think `sendOrFail` is enough in my opinion.